cranelift/frontend/src/ssa.rs

```diff
@@ -545,76 +544,47 @@ impl SSABuilder {
             // There is disagreement in the predecessors on which value to use so we have
             // to keep the block argument.
             let mut preds = self.ssa_blocks[dest_block].predecessors;
-            let var_defs = &mut self.variables[var];
             for (idx, &val) in results.as_slice().iter().enumerate() {
                 let pred = preds.get_mut(idx, &mut self.inst_pool).unwrap();
                 let branch = *pred;
-                if let Some((new_block, new_branch)) =
-                    Self::append_jump_argument(func, branch, dest_block, val)
-                {
-                    *pred = new_branch;
-                    let old_block = func.layout.inst_block(branch).unwrap();
-                    self.ssa_blocks[new_block] = SSABlockData {
-                        predecessors: EntityList::from_slice(&[branch], &mut self.inst_pool),
-                        sealed: Sealed::Yes,
-                        single_predecessor: PackedOption::from(old_block),
-                    };
-                    var_defs[new_block] = PackedOption::from(val);
-                    self.side_effects.split_blocks_created.push(new_block);
-                }
+                Self::append_jump_argument(func, branch, dest_block, val)
             }
             sentinel
         }
     }
 
     /// Appends a jump argument to a jump instruction, returns block created in case of
     /// critical edge splitting.
-    fn append_jump_argument(
-        func: &mut Function,
-        branch: Inst,
-        dest_block: Block,
-        val: Value,
-    ) -> Option<(Block, Inst)> {
+    fn append_jump_argument(func: &mut Function, branch: Inst, dest_block: Block, val: Value) {
         let dfg = &mut func.stencil.dfg;
         match &mut dfg.insts[branch] {
             // For a single destination appending a jump argument to the instruction
             // is sufficient.
-            InstructionData::Jump { destination, .. } => {
-                destination.append_argument(val, &mut dfg.value_lists);
-                None
+            InstructionData::Jump {
+                destination: dest, ..
+            } => {
+                dest.append_argument(val, &mut dfg.value_lists);
             }
             InstructionData::Brif { blocks, .. } => {
                 for block in blocks {
                     if block.block(&dfg.value_lists) == dest_block {
                         block.append_argument(val, &mut dfg.value_lists);
                     }
                 }
-                None
             }
             InstructionData::BranchTable {
-                table: jt,
-                destination,
-                ..
+                table, destination, ..
             } => {
-                // In the case of a jump table, the situation is tricky because br_table doesn't
-                // support arguments. We have to split the critical edge.
```

Comment:
This code may need to be preserved for `invoke` depending on how it will be implemented. The exception path gets some block arguments from the personality function, so either those would be added to the block arguments listed in `BlockCall` of the `invoke` instruction, or the `invoke` instruction would only list a `Block` and synthesize the full block argument list internally. What do you think would be cleaner?

Reply:
I like the idea of  using a `BlockCall` on the `invoke` instruction. I think the path for doing that should be pretty well worn after adding `brif`, but if you run into any problems using `BlockCall` I'd be happy to help work through them :)

Reply:
The difference between `brif` and `invoke` is that `brif` passes all block arguments in the instruction to the destination block, but for `invoke` there would be some extra block arguments not specified in the `BlockCall`. Do you add those at the start or the end? And is there any code that expects there to be no block arguments added? The following is pseudocode to demonstrate the issue:

```
block0:
    v1 = iconst.i64 42
    invoke fn0(), block1, block2(v1)

block1:
    trap user0

block2(v2: i64, v3: i64):
    ; this is the landingpad
    ; either v2 or v3 is the extra argument containing a value passed by the personality function. which one should be the extra argument?
    trap user0
```

Reply:
Ah, interesting! I would add them to the end and still use a `BlockCall`. It's sort of like a partially applied function that way.

Reply:
Sure, will try that once I get to adding the invoke instruction.